doc/man/man3/DwmFileIO.hh.3
.TH "DwmFileIO.hh" 3 "libDwm-0.0.20240716" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DwmFileIO.hh \- \fBDwm::FileIO\fP class declaration\&.  

.SH SYNOPSIS
.br
.PP
\fR#include <array>\fP
.br
\fR#include <cassert>\fP
.br
\fR#include <cstdint>\fP
.br
\fR#include <cstdio>\fP
.br
\fR#include <deque>\fP
.br
\fR#include <iostream>\fP
.br
\fR#include <list>\fP
.br
\fR#include <map>\fP
.br
\fR#include <set>\fP
.br
\fR#include <string>\fP
.br
\fR#include <tuple>\fP
.br
\fR#include <unordered_map>\fP
.br
\fR#include <unordered_set>\fP
.br
\fR#include <variant>\fP
.br
\fR#include <vector>\fP
.br
\fR#include 'DwmPortability\&.hh'\fP
.br
\fR#include 'DwmFileIOCapable\&.hh'\fP
.br
\fR#include 'DwmVariantFromIndex\&.hh'\fP
.br

.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBDwm::FileIO\fP"
.br
.RI "This class contains a collection of static functions for reading and writing simple types, in network byte order (MSB first)\&. "
.in -1c
.SH "Detailed Description"
.PP 
\fBDwm::FileIO\fP class declaration\&. 


.SH "Author"
.PP 
Generated automatically by Doxygen for libDwm-0\&.0\&.20240716 from the source code\&.


ChangeLog
2024-07-25  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as libDwm-0.9.45

	* README.md: mention libfmt dependency for the case where
	std::format is not available.

	* tagged as libDwm-0.9.44

	* classes/include/DwmGZIOCapable.hh: Documentation updates.
	* classes/include/DwmBZ2IOCapable.hh: Documentation updates.
	* classes/include/DwmASIOCapable.hh: Documentation updates.
	* classes/include/DwmDescriptorIOCapable.hh: Documentation updates.
	* classes/include/DwmFileIOCapable.hh: Documentation updates.
	* classes/include/DwmStreamIOCapable.hh: Documentation updates.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmDocMainPage.hh, README.md,
	doc/examples/DwmIOExample3.cc: Documentation updates.

2024-07-24  dwm  <dwm@nomnom.mcplex.net>

	* tagged as libDwm-0.9.43

	* mkrelease.sh: 'push --tags' should be 'push --follow-tags'

	* tagged as libDwm-0.9.42

	* classes/include/DwmSysLogger.hh: Fix pasto on platforms where
	we need libfmt (std:: should be fmt::), namely Raspberry Pi OS.

2024-07-24  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.41

	* classes/tests/TestThreadQueue.cc: Use max concurrency on
	multicore systems when running performance tests, just to
	add to the stress.  Also use C++14 digit separator (') in
	some numeric literals for readability.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/src/DwmSysLogger.cc, classes/include/DwmSysLogger.hh:
	Allow use of std::source_location, and use them in the Syslog
	and FSyslog macros.

	* Removed the following files since I don't need them here:

	classes/include/DwmHostPinger.hh
	classes/include/DwmIpv4IcmpDestinationUnreachable.hh
	classes/include/DwmIpv4IcmpEchoReply.hh
	classes/include/DwmIpv4IcmpEchoRequest.hh
	classes/include/DwmIpv4IcmpMessage.hh
	classes/include/DwmIpv4IcmpRedirect.hh
	classes/include/DwmIpv4IcmpSourceQuench.hh
	classes/include/DwmIpv4IcmpTimeExceeded.hh
	classes/include/DwmIpv4PacketHeader.hh
	classes/include/DwmIpv4TcpHeader.hh
	classes/include/DwmIpv4TcpPacket.hh
	classes/include/DwmIpv4TcpPayload.hh
	classes/include/DwmIpv4UdpHeader.hh
	classes/include/DwmIpv4UdpPacket.hh
	classes/include/DwmIpv4UdpPayload.hh
	classes/include/DwmPatricia.hh
	classes/include/DwmPingDestination.hh
	classes/include/DwmPingDestinationEntry.hh
	classes/include/DwmSocket.hh
	classes/include/DwmXmlAttribute.hh
	classes/include/DwmXmlElement.hh
	classes/src/DwmHostPinger.cc
	classes/src/DwmIpv4IcmpDestinationUnreachable.cc
	classes/src/DwmIpv4IcmpEchoReply.cc
	classes/src/DwmIpv4IcmpEchoRequest.cc
	classes/src/DwmIpv4IcmpMessage.cc
	classes/src/DwmIpv4IcmpRedirect.cc
	classes/src/DwmIpv4IcmpSourceQuench.cc
	classes/src/DwmIpv4IcmpTimeExceeded.cc
	classes/src/DwmIpv4PacketHeader.cc
	classes/src/DwmIpv4TcpHeader.cc
	classes/src/DwmIpv4TcpPacket.cc
	classes/src/DwmIpv4TcpPayload.cc
	classes/src/DwmIpv4UdpHeader.cc
	classes/src/DwmIpv4UdpPacket.cc
	classes/src/DwmIpv4UdpPayload.cc
	classes/src/DwmPingDestination.cc
	classes/src/DwmPingDestinationEntry.cc
	classes/src/DwmSocket.cc
	classes/src/DwmXmlAttribute.cc
	classes/src/DwmXmlElement.cc
	classes/src/Makefile
	classes/tests/Makefile
	classes/tests/TestDwmSocket.cc
	classes/tests/TestHostPinger.cc
	classes/tests/TestIpv4IcmpDestinationUnreachable.cc
	classes/tests/TestIpv4IcmpEchoReply.cc
	classes/tests/TestIpv4IcmpEchoRequest.cc
	classes/tests/TestIpv4IcmpRedirect.cc
	classes/tests/TestIpv4IcmpSourceQuench.cc
	classes/tests/TestIpv4IcmpTimeExceeded.cc
	classes/tests/TestIpv4PacketHeader.cc
	classes/tests/TestIpv4TcpHeader.cc
	classes/tests/TestIpv4TcpPacket.cc
	classes/tests/TestIpv4UdpHeader.cc
	classes/tests/TestIpv4UdpPacket.cc
	classes/tests/TestPatricia.cc
	classes/tests/ip.pcap
	classes/tests/tcp.pcap
	classes/tests/udp.pcap

2024-07-23  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* getvers.sh: bump shared library version to reflect ABI changes
	(vtables changed from removal of inheritance).

	* README.md: Document Read and Write requirements for
	ASIO with boost::asio::generic::stream_protocol::socket

	* classes/include/DwmIpv4IcmpMessage.hh,
	  classes/include/DwmIpv4PacketHeader.hh,
	  classes/include/DwmIpv4TcpHeader.hh,
	  classes/include/DwmIpv4TcpPacket.hh,
	  classes/include/DwmIpv4TcpPayload.hh,
	  classes/include/DwmIpv4UdpHeader.hh,
	  classes/include/DwmIpv4UdpPacket.hh,
	  classes/include/DwmIpv4UdpPayload.hh:
	No longer need to inherit from *Readable and *Writable classes.

	* classes/include/DwmASIOCapable.hh,
	  classes/include/DwmBZ2IOCapable.hh,
	  classes/include/DwmDescriptorIOCapable.hh,
	  classes/include/DwmFileIOCapable.hh,
	  classes/include/DwmGZIOCapable.hh,
	  classes/include/DwmStreamIOCapable.hh:
	Minor cleanup to concepts.

	* tagged as libDwm-0.9.40

	* doc/.gitignore: ignore libDwm.tag (generated by doxygen)

	* README.md: Add mkfbsdmnfst and mkdebcontrol as build
	dependencies.

2024-07-22  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* Removed IN-PROGRESS file.

	* tagged as libDwm-0.9.39

	* Added README.md for the github mirror.

2024-07-21  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.38

2024-07-21  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* doc/footer.html: Tweak to point at my web site.

	* classes/src/DwmMacAddress.cc: Use sizeof() where appropriate.

	* Remove doxygen-generated documentation from repository; it's
	build output that doesn't need to be in the repository.

2024-07-20  Daniel McRobb  <dwm@toto.mcplex.net>

	* tagged as libDwm-0.9.37

2024-07-20  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* classes/src/DwmProcessTableLex.lex,
	classes/src/DwmProcessTableParse.y: 'ps' output on FreeBSD shows
	a single hyphen ('-') instead of a ctime() string for lstart for
	some processes.  Patch this by substituting the lstart from the
	nearest ancestor of the process that has a valid lstart.

	* classes/include/DwmBootTime.hh, classes/include/DwmUptime.hh
	classes/src/DwmBootTime.cc, classes/src/DwmUptime.cc:

classes/src/DwmFileIO.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2004-2007, 2013, 2017, 2020
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file DwmFileIO.cc
//!  \author Daniel W. McRobb
//!  \brief Dwm::FileIO class implementation
//---------------------------------------------------------------------------

#include <cstdlib>
#include <cassert>

#include "DwmSysLogger.hh"
#include "DwmFileIO.hh"
#include "DwmPortability.hh"
#include "DwmXDRUtils.hh"

namespace Dwm {

  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, char c)
  {
    size_t  rc = 0;
    if (f) {
      rc = fwrite((void *)&c, sizeof(c), 1, f);
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, uint8_t c)
  {
    size_t  rc = 0;
    if (f) {
      rc = fwrite((void *)&c, sizeof(c), 1, f);
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, bool b)
  {
    uint8_t  c = b;
    return(FileIO::Write(f, c));
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, int16_t val)
  {
    size_t rc = 0;
    if (f) {
      int16_t v = htobe16(val);
      if (fwrite((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
      }
    }
    return(rc);
  }
   
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, uint16_t val)
  {
    size_t rc = 0;
    if (f) {
      uint16_t v = htobe16(val);
      if (fwrite((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, int32_t val)
  {
    size_t rc = 0;
    if (f) {
      int32_t v = htobe32(val);
      if (fwrite((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, uint32_t val)
  {
    size_t rc = 0;
    if (f) {
      uint32_t v = htobe32(val);
      if (fwrite((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, const int64_t & val)
  {
    size_t  rc = 0;
    if (f) {
      int64_t  v = htobe64(val);
      if (fwrite((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, const uint64_t & val)
  {
    size_t  rc = 0;
    if (f) {
      uint64_t  v = htobe64(val);
      if (fwrite((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
      }
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, float val)
  {
    size_t  rc = 0;
    if (f) {
      char  *buf = 0;
      rc = XDRUtils::Encode(val, &buf);
      if (fwrite((void *)buf, rc, 1, f) != 1) {
        rc = 0;
      }
      free(buf);
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, const double & val)
  {
    size_t  rc = 0;
    
    if (f) {
      char  *buf = 0;
      rc = XDRUtils::Encode(val, &buf);
      if (fwrite((void *)buf, rc, 1, f) != 1) {
        rc = 0;
      }
      free(buf);
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Write(FILE * f, const std::string & s)
  {
    size_t  rc = 0;
    if (f) {
      uint64_t len = s.size();
      if (Write(f, len) == sizeof(len)) {
        rc = sizeof(len);
        if (len > 0) {
          if (fwrite((void *)s.c_str(), len, 1, f) == 1) {
            rc += len;
          }
          else {
            rc = 0;
          }
        }
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, char & c)
  {
    size_t rc = 0;
    if (f) {
      rc = fread((void *)&c, sizeof(c), 1, f);
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, uint8_t & c)
  {
    size_t  rc = 0;
    if (f) {
      rc = fread((void *)&c, sizeof(c), 1, f);
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, bool & b)
  {
    uint8_t  c;
    size_t  rc = FileIO::Read(f, c);
    if (rc > 0) {
      b = c ? true : false;
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, int16_t & val)
  {
    size_t  rc = 0;
    if (f) {
      int16_t v;
      if (fread((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
        val = be16toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, uint16_t & val)
  {
    size_t  rc = 0;
    if (f) {
      uint16_t v;
      if (fread((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
        val = be16toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, int32_t & val)
  {
    size_t  rc = 0;
    if (f) {
      int32_t v;
      if (fread((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
        val = be32toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, uint32_t & val)
  {
    size_t  rc = 0;
    if (f) {
      uint32_t v;
      if (fread((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
        val = be32toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, int64_t & val)
  {
    size_t  rc = 0;
    if (f) {
      uint64_t  v;
      if (fread((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
        val = be64toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, uint64_t & val)
  {
    size_t  rc = 0;
    if (f) {
      uint64_t v;
      if (fread((void *)&v, sizeof(v), 1, f) == 1) {
        rc = sizeof(v);
        val = be64toh(v);
      }
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, float & val)
  {
    size_t  rc = 0;
    
    if (f) {
      char   buf[4];
      if (fread((void *)buf, 4, 1, f) == 1) {
        XDRUtils::Decode(buf, val);
        rc = 4;
      }
    }
    
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, double & val)
  {
    size_t  rc = 0;
    
    if (f) {
      char   buf[8];
      if (fread((void *)buf, 8, 1, f) == 1) {
        XDRUtils::Decode(buf, val);
        rc = 8;
      }
    }

    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  size_t FileIO::Read(FILE * f, std::string & s)
  {
    s.clear();
    size_t  rc = 0;
    if (f) {
      uint64_t len;
      if (Read(f, len) == sizeof(len)) {
        rc = sizeof(len);
        if (len > 0) {
          try {
            s.resize(len);
            if (fread((void *)s.data(), len , 1, f) == 1) {
              rc += len;
            }
            else {
              s.clear();
              rc = 0;
            }
          }
          catch (const std::exception & ex) {
            Syslog(LOG_ERR, "Exception: %s", ex.what());
            s.clear();
            rc = 0;
          }
          catch (...) {
            Syslog(LOG_ERR, "Exception");
            s.clear();
            rc = 0;
          }
        }
      }
    }
    return(rc);
  }

}  // namespace Dwm


//---------------------------- emacs settings -----------------------------
//  Local Variables:
//  mode: C++
//  tab-width: 2
//  indent-tabs-mode: nil
//  c-basic-offset: 2
//  End:
//-------------------------------------------------------------------------
